Add global --[no-]color flags

iliekturtles opened the pull request:
First pass ready for review! I went with the option of always attempting to create a color shell and then checking the `color_when` and `tty` flags to determine if `Shell` should actually write color output.

Should I squash the commits?

rust-highfive commented:
r? @huonw

(rust_highfive has picked a reviewer for you, use r? to override)

huonw commented:
Hm I think this sort of flag is typically done as three states: no colours, auto-detect and always colour, e.g. `rustc` exposes the three options as arguments to a `--color` flag: `--color never`, `--color auto` and `--color always`.

iliekturtles commented:
I had modeled the two flags after the dual `--verbose|--quiet` flags. I'll go back and switch to a `--color=always|auto|never`.

iliekturtles commented:
@huonw `--color [auto|always|never]` implementation ready for review. To get docopt to parse the option properly I needed to do `--color WHEN` which means just `--color` isn't valid. I think that using `--color=WHEN` would allow for an empty option, but that is inconsistent with other cargo options.

alexcrichton commented on src/cargo/core/shell.rs:
```diff
@@ -115,6 +122,19 @@ impl MultiShell {
+
+            None => Auto,
+
+            Some(_) => return Err(human("argument for --color must be auto, always, or never")),
```
Can this error also mention what the provided color was? Something like:

```
expected argument for --color to be auto, always, or never, but found `{}`
```

alexcrichton commented on src/cargo/core/shell.rs:
```diff
@@ -18,10 +19,16 @@ pub enum Verbosity {
 #[derive(Clone, Copy)]
 pub struct ShellConfig {
-    pub color: bool,
-    pub verbosity: Verbosity,
```
I think this may have been accidentally deleted?

iliekturtles replied:
When I first started investigating the problem I discovered that both `Shell` and `MultiShell` had verbosity options. The option in `Shell` always used it's default value and was never changed. I removed the option on my path to the first implementation for this PR and the original commit ended up squashed. Let me know if you would like to see it returned to keep this PR cleaner.

alexcrichton replied:
Ah no this is fine in that case, thanks for the cleanup!

alexcrichton commented on tests/test_shell.rs:
```diff
@@ -60,17 +60,20 @@ test!(colored_shell {
 });
 
-test!(quiet_shell {
-    let config = ShellConfig { color: true, verbosity: Quiet, tty: true };
```
Looks like this test was accidentally deleted?

alexcrichton replied:
(test was still deleted)

iliekturtles replied:
Same as the previous comment. `Shell` no longer has a concept of verbosity so this test didn't have anything left to test.

alexcrichton commented on src/cargo/core/shell.rs:
```diff
@@ -182,30 +179,42 @@ impl Shell {
-            Colored(ref mut c) => c.fg(color),
-            NoColor(_) => Ok(false)
+        let color_config = self.config.tty && Auto == self.config.color_config
+            || Always == self.config.color_config;
```
Can you refactor this logic into a helper function? That way it can just be:

```rust
match self.terminal {
    Colored(ref mut c) if self.colored() => c.fg(color),
    _ => Ok(false)
}
```

alexcrichton commented on src/cargo/core/shell.rs:
```diff
@@ -122,49 +142,26 @@ impl MultiShell {
-                try!(self.flush());
-                Ok(())
-            },
-        }
```
Looks like this merge went awry

alexcrichton replied:
(I think this still needs to be udpated)

iliekturtles replied:
Same here. No verbosity so the match was removed and the contents of the `_` arm were left. This was the original functionality.

alexcrichton commented on src/bin/bench.rs:
```diff
@@ -43,6 +44,9 @@ Options:
     -q, --quiet              No output printed to stdout
+    --color WHEN             Configure coloring of output: auto = colorize, if
+                             output goes to a tty (default); always = always
+                             colorize output; never = never colorize output
```
Can this description be a little more concise? It'd be nice to keep it on one line

alexcrichton commented:
Thanks @iliekturtles! Looks like there's a number of failing tests as well which will need to be addressed.

bors commented:
:umbrella: The latest upstream changes (presumably #1835) made this pull request unmergeable. Please resolve the merge conflicts.

iliekturtles commented:
@alexcrichton Changes noted above resolved. I will get the required dependencies to build on Windows installed this week to investigate the Windows build failures.

* I couldn't put the call to `self.colored()` in the match guard because `self.terminal` is borrowed as mutable.
 * Is the `--color WHEN` help message too concise now?
 * Rebased but didn't squash. Is this the appropriate why to update the PR? I figured a merge from master wasn't desirable.

alexcrichton commented on src/cargo/core/shell.rs:
```diff
@@ -115,6 +122,21 @@ impl MultiShell {
+
+            Some(arg) => return Err(human(
+                    format!("argument for --color must be auto, always, or never, but found `{}`",
+                           arg))),
```
This is typically formatted as:

```
Some(arg) => {
    return Err(human(format!("foo \
                              bar \
                              baz"
                             foo)))
}
```

alexcrichton commented:
> I couldn't put the call to self.colored() in the match guard because self.terminal is borrowed as mutable.

Ah that's fine

> Is the --color WHEN help message too concise now?

Looks good to me!

> Rebased but didn't squash. Is this the appropriate why to update the PR? I figured a merge from master wasn't desirable.

That's ok, you can squash once it's ready to go.

iliekturtles commented:
@alexcrichton Formatting fixed. I couldn't get Cargo compiling on Windows yet (libssh2 fails) but looking at the logs and thinking about the error I believe I may know what the issue is.

I remember reading somewhere (from you perhaps) that color output on Windows requires changing the console's configuration rather than just marker bytes in the output stream. The `tty` check in `Shell::create` was removed to allow for `--color always` and the `TerminfoTerminal::new(out)` call is likely failing on Windows and sending all output to stderr.

`--color always` may not always be possible on Windows and I've pushed the change for you and the buildbots to review and will see if I can make progress compiling libssh2 later this week.

alexcrichton commented on src/etc/cargo.1:
```diff
@@ -26,6 +26,9 @@ List all available cargo commands
 \fB\-v, \-\-verbose\fR
 Use verbose output
+.TP
+\fB\-v, \-\-color\fR
```
the `-v` header here can be removed (it's just `--color`)

alexcrichton commented on src/cargo/core/shell.rs:
```diff
@@ -122,7 +144,7 @@ impl MultiShell {
 impl Shell {
     pub fn create(out: Box<Write + Send>, config: ShellConfig) -> Shell {
-        if config.tty && config.color {
+        if !cfg!(windows) || config.tty {
```
Could you add a comment explaining why this `cfg!` is here?

alexcrichton commented:
Ok, looks good to me, just a few last nits! r=me with those + a squash

iliekturtles commented:
Squashed, rebased, and fixed nits. Thanks for all the reviews!

alexcrichton commented:
@bors: r+ 87053d4b763b649413b67e299d8dc420a2d4b357

bors commented:
:hourglass: Testing commit 87053d4 with merge a475c8c...

bors commented:
:broken_heart: Test failed - [cargo-linux-64](http://buildbot.rust-lang.org/builders/cargo-linux-64/builds/1223)

iliekturtles commented:
@alexcrichton Any idea what's happening here? I can't reproduce on my Linux x86-64 machine and the travis-ci build previously succeeded. The errors almost look like the Windows failures we were seeing where `TerminfoTerminal::new(out)` was failing and all output was going to stderr.

alexcrichton commented on src/cargo/core/shell.rs:
```diff
@@ -122,7 +144,9 @@ impl MultiShell {
-        if config.tty && config.color {
+        // Check for cfg!(windows) as colored output on Windows can only be supported when a tty is
+        // present.
+        if !cfg!(windows) || config.tty {
```
Perhaps this changed line is the one causing failures? I'm also not 100% sure why the bots are failing

iliekturtles commented:
Did some digging and figured out how to reproduce locally by clearing `$TERM` when running `make test`. Committed a refactored `Shell::create` that inspects `TermInfo` availability before creating a `TerminfoTerminal`.

I'm looking for some advice on creating a test to explicitly capture this for test bots where `$TERM` is set. `evn::remove_var` and `env::set_var` don't seem like great options.

alexcrichton commented:
Isn't this the same as before? It looks like you've just inlined the logic of `TerminfoTerminal::new` directly into this function.

You should be able to modify the environment variables of spawned processes such that they don't affect the parent which should help in writing at est.

iliekturtles commented:
Not quite the same as before. The previous code using `TerminfoTerminal::new` consumed `out` so if the constructor failed the only possible solution was to fall back to `io::stderr`. The new code matches `TermInfo::from_env` first and then determines what kind of shell to create. This catches the Windows specific work-around and the failures we saw on the cargo-linux-64 build-bots.

Am I understanding correct that each `#test` is run as a separate process? I rebased and committed a new test, `no_term`, under this assumption to validate output when a `TerminfoTerminal` cannot be created.

alexcrichton commented:
Ah I see what you mean by the difference now. The entire shell creation step seems super confusing and convoluted now, and it may be worth some reorganization at some point... 

Unfortunately each test isn't run in a separate process, so this is a bit racy as-is. You'll have to manually spawn a new process (e.g. `cargo`) and then check for colorization afterwards.

iliekturtles commented:
Rebased to get your Windows tests fixes and pushed a new commit which runs a separate Cargo process for the `no_term` test. Also added some comments to shell creation to explain the logic. I may be too immersed to find any simplifications.

alexcrichton commented:
@bors: r+ bcf2a11da445fdd71af8eb9eeb5591816571b439

Thanks!

bors commented:
:hourglass: Testing commit bcf2a11 with merge 03d3c85...

bors commented:
:sunny: Test successful - [cargo-linux-32](http://buildbot.rust-lang.org/builders/cargo-linux-32/builds/1242), [cargo-linux-64](http://buildbot.rust-lang.org/builders/cargo-linux-64/builds/1232), [cargo-mac-32](http://buildbot.rust-lang.org/builders/cargo-mac-32/builds/1239), [cargo-mac-64](http://buildbot.rust-lang.org/builders/cargo-mac-64/builds/1235), [cargo-win-gnu-32](http://buildbot.rust-lang.org/builders/cargo-win-gnu-32/builds/113), [cargo-win-gnu-64](http://buildbot.rust-lang.org/builders/cargo-win-gnu-64/builds/114), [cargo-win-msvc-32](http://buildbot.rust-lang.org/builders/cargo-win-msvc-32/builds/31), [cargo-win-msvc-64](http://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/58)
